cmds/strings/strings.go

```diff
@@ -34,50 +34,34 @@ func asciiIsPrint(char byte) bool {
 	return char >= 32 && char <= 126
 }
 
-func stringsIO(r io.Reader, w io.Writer) error {
-	// At least `n` bytes must be held in memory at a given time.
-	rb := bufio.NewReaderSize(r, *n)
-
-	// This processes the file one byte at a time. It might be inefficient,
-	// but it works for now.
-outerLoop:
+func stringsIO(r *bufio.Reader, w io.Writer) error {
+	var o []byte
 	for {
-		// Discard bytes from the buffer until the first `n` bytes are
-		// all printable.
-		peek, err := rb.Peek(*n)
+		b, err := r.ReadByte()
 		if err == io.EOF {
+			if len(o) >= *n {
+				w.Write(o)
+				w.Write([]byte{'\n'})
+			}
 			return nil
 		}
 		if err != nil {
 			return err
 		}
-		for i := *n - 1; i >= 0; i-- {
-			if !asciiIsPrint(peek[i]) {
-				rb.Discard(i + 1)
-				continue outerLoop
-			}
-		}
-
-		// Write the first `n` bytes of the buffer.
-		w.Write(peek)
-		rb.Discard(*n)
-
-		// Keep writing bytes until a non-printable byte is encountered.
-		for {
-			b, err := rb.ReadByte()
-			if err == io.EOF {
+		if !asciiIsPrint(b) {
+			if len(o) >= *n {
+				w.Write(o)
 				w.Write([]byte{'\n'})
-				return nil
-			}
-			if err != nil {
-				return err
 			}
-			if !asciiIsPrint(b) {
-				w.Write([]byte{'\n'})
-				break
-			}
-			w.Write([]byte{b})
+			o = o[:0]
+			continue
+		}
+		// Prevent the buffer from growing indefinitely.
+		if len(o) >= *n+1024 {
+			w.Write(o[:1024])
+			o = o[1024:]
```

Comment:
if you use the make([]byte, 0, 4096) could this be
o = o[0:0]?
https://play.golang.org/p/8pkiD_P3TP2

Reply:
This can't be `o = o[0:0]` because it would change the condition `if len(o) >= *n {` on line 52. Essentially, if `o` contains more than `k + n` bytes, we can pop the `k`-length prefix of `o`.